solutions/cf-1528A.mdx

```diff
@@ -0,0 +1,110 @@
+---
+id: cf-1528A
+source: CF
+title: Parsa's Humongous Tree
+author: Ryan Chou
+---
+
+[Official Editorial (C++)](https://codeforces.com/blog/entry/91058)
+
+## Explanation
+
+Let's go through a case where we've already computed the beauty values for all of vertex $u$'s neighbors.
+
+Let $\texttt{range}_u$ store $l_u$ and $r_u$ and $\texttt{adj}_u$ store the beauty values of all vertices adjacent to $u$.
+
+$\texttt{range}_u = \{1, 10\}$
+
+$\texttt{adj}_u = \{2, 3, 6, 9\}$
+
+Let's say that we select $7$, can we do better?
+
+The current beauty sum is: $|7-2| + |7-3| + |7-6| + |7-9| = 12$.
+
+Can we do better?
+
+Intuitively, for any $X$ that we choose to be $u$'s beauty value, if there's more elements that are greater than this $X$, making $X$ as small as possible would be more optimal (increasing differences for these elements) and vice versa. Note that in a case where the number of greater beauty values is the same as the number of smaller beauty values, setting $X$ to the minimum or maximum will never decrease the answer.
+
+It turns out, this observation is all we need. Since setting $u$ to $l_u$ or $r_u$ is always optimal, we'll handle this through DP.
+
+If $\texttt{dp}[i][j]$ stores the maximum sum of beauty values at the subtree rooted at $i$ when $i$'s "beauty type" is $j$. Since $i$ can only take two values, $j$ will be equal to $0$ when $u$'s beauty value is $l_u$ and $1$ when $u$'s beauty value is $r_u$.
+
+Our transitions are:
+
+$\texttt{dp}_{u \space 0} \mathrel{+}= \max(\texttt{dp}_{v \space 0} + |range_{u \space 0} - range_{v \space 0}|, \texttt{dp}_{v \space 1} + |range_{u \space 1} - range_{v \space 0}|)$
+
+$\texttt{dp}_{u \space 1} \mathrel{+}= \max( \texttt{dp}_{v \space 1} + |range_{u \space 1} - range_{v \space 1}|, \texttt{dp}_{v \space 0} + |range_{u \space 0} - range_{v \space 1}|)$
```

Comment:
(optional) explain this with plain text instead of some eldritch math formula

Reply:
i think ill use both